### PR TITLE
#129: mark job status updates for non-existent job steps as propagated …

### PR DESCRIPTION
**Pull request by slr71**
…if there are too many or too few matching job steps for processing

### CODE REVIEW

File: src/apps/service/apps.clj [hunk 1]
```diff
@@ -219,17 +219,18 @@
...
 (defn update-job-status
   ([external-id]
+   (jobs/validate-job-status-update-step-count external-id)
    (transaction
-    (let [{job-id :job_id} (jobs/get-unique-job-step external-id)
-          updates          (jp/get-job-status-updates external-id)]
+    (let [updates (jp/get-job-status-updates external-id)]
       (when (seq updates)
-        (let [job-step    (jobs/lock-job-step job-id external-id)
+        (let [job-id      (:job_id (jp/get-job-steps-by-external-id external-id))
```
Comment (psarando) [inline src/apps/service/apps.clj]: It seems like `get-job-steps-by-external-id` will return a list of results?
Should `first` be called on the results before `:job_id`?

File: src/apps/service/apps/jobs.clj [hunk 2]
```diff
@@ -14,16 +14,18 @@
...
-  "Gets a unique job step for an external ID. An exception is thrown if no job step
-  is found or if multiple job steps are found."
+(defn validate-job-status-update-step-count
+  "Validates the number of steps for an external ID provided in a set of job status updates. If there are too many or
+  too few matching job steps then all of the job status updates will be marked as propagated and an exception will be
+  thrown. This function should not be called from within a transaction."
   [external-id]
   (let [job-steps (jp/get-job-steps-by-external-id external-id)]
     (when (empty? job-steps)
+      (jp/mark-job-status-updates-for-external-id-completed external-id)
```
Comment (psarando) [inline src/apps/service/apps/jobs.clj]: 👍

File: src/apps/service/apps.clj (hunk 1 above)
Comment (slr71) [reply src/apps/service/apps.clj]: Ah, yes it should. Thanks for catching that.

**Comment by slr71** [comment]
Thanks for the review!

**Pull request merged**